File: src/main/java/com/humanizar/units/mapper/UnitMapper.java
package com.humanizar.units.mapper;

import java.util.UUID;

import org.springframework.stereotype.Component;

import com.humanizar.units.dto.UnitDTO;
import com.humanizar.units.exception.UnitException;
import com.humanizar.units.model.Units;
import com.humanizar.units.model.enums.ReasonCode;

@Component
public class UnitMapper {

    public void validate(UnitDTO unitDTO) {
        if (unitDTO == null) {
            throw new UnitException(ReasonCode.VALIDATION_ERROR, "Unidade nao pode ser nula.");
        }

        validateRequired("unitName", unitDTO.unitName());
        validateRequired("razaoSocial", unitDTO.razaoSocial());
        validateRequired("endereco", unitDTO.endereco());
        validateRequired("numero", unitDTO.numero());
        validateRequired("bairro", unitDTO.bairro());
        validateRequired("cep", unitDTO.cep());
        validateRequired("cnpj", unitDTO.cnpj());
    }

    public Units toEntity(UUID municipioId, UnitDTO unitDTO) {
        validate(unitDTO);

        return Units.builder()
                .id(unitDTO.unitId())
                .municipioId(municipioId)
                .unitName(unitDTO.unitName())
                .razaoSocial(unitDTO.razaoSocial())
                .endereco(unitDTO.endereco())
                .numero(unitDTO.numero())
                .complemento(unitDTO.complemento())
                .bairro(unitDTO.bairro())
                .cep(unitDTO.cep())
                .cnpj(unitDTO.cnpj())
                .build();
    }

    public UnitDTO toDTO(Units entity) {
        if (entity == null) {
            throw new UnitException(ReasonCode.PERSISTENCE_FAILURE, "Entidade de unidade nao pode ser nula.");
        }

        return new UnitDTO(
                entity.getId(),
                entity.getMunicipioId(),
                entity.getUnitName(),
                entity.getRazaoSocial(),
                entity.getEndereco(),
                entity.getNumero(),
                entity.getComplemento(),
                entity.getBairro(),
                entity.getCep(),
                entity.getCnpj());
    }

    public void applyUpdates(Units entity, UnitDTO unitDTO) {
        validate(unitDTO);

        entity.setUnitName(unitDTO.unitName());
        entity.setRazaoSocial(unitDTO.razaoSocial());
        entity.setEndereco(unitDTO.endereco());
        entity.setNumero(unitDTO.numero());
        entity.setComplemento(unitDTO.complemento());
        entity.setBairro(unitDTO.bairro());
        entity.setCep(unitDTO.cep());
        entity.setCnpj(unitDTO.cnpj());
    }

    private void validateRequired(String fieldName, String value) {
        if (value == null || value.isBlank()) {
            throw new UnitException(
                    ReasonCode.VALIDATION_ERROR,
                    "Campo obrigatorio invalido: " + fieldName + ".");
        }
    }
}

File: src/application/mapper/unit_mapper.rs
use uuid::Uuid;

use crate::application::dto::UnitDto;
use crate::domain::exception::UnitException;
use crate::domain::model::Unit;

use super::field_validator::required_field;

pub struct UnitMapper;

impl UnitMapper {
    pub fn to_new_domain(municipio_id: Uuid, dto: &UnitDto) -> Result<Unit, UnitException> {
        let fields = validated_fields(dto)?;

        Ok(Unit::new(
            municipio_id,
            fields.unit_name,
            fields.razao_social,
            fields.endereco,
            fields.numero,
            dto.complemento().map(str::to_owned),
            fields.bairro,
            fields.cep,
            fields.cnpj,
        ))
    }

    pub fn to_updated_domain(current: &Unit, dto: &UnitDto) -> Result<Unit, UnitException> {
        let fields = validated_fields(dto)?;

        Ok(Unit::restore(
            current.id(),
            current.municipio_id(),
            fields.unit_name,
            fields.razao_social,
            fields.endereco,
            fields.numero,
            dto.complemento().map(str::to_owned),
            fields.bairro,
            fields.cep,
            fields.cnpj,
            current.created_at(),
            current.updated_at(),
        ))
    }

    pub fn to_dto(unit: Unit) -> UnitDto {
        UnitDto::new(
            unit.id(),
            Some(unit.municipio_id()),
            Some(unit.unit_name().to_owned()),
            Some(unit.razao_social().to_owned()),
            Some(unit.endereco().to_owned()),
            Some(unit.numero().to_owned()),
            unit.complemento().map(str::to_owned),
            Some(unit.bairro().to_owned()),
            Some(unit.cep().to_owned()),
            Some(unit.cnpj().to_owned()),
        )
    }

    pub fn to_dtos(units: Vec<Unit>) -> Vec<UnitDto> {
        units.into_iter().map(Self::to_dto).collect()
    }
}

struct UnitFields {
    unit_name: String,
    razao_social: String,
    endereco: String,
    numero: String,
    bairro: String,
    cep: String,
    cnpj: String,
}

fn validated_fields(dto: &UnitDto) -> Result<UnitFields, UnitException> {
    Ok(UnitFields {
        unit_name: required_field("unitName", dto.unit_name())?,
        razao_social: required_field("razaoSocial", dto.razao_social())?,
        endereco: required_field("endereco", dto.endereco())?,
        numero: required_field("numero", dto.numero())?,
        bairro: required_field("bairro", dto.bairro())?,
        cep: required_field("cep", dto.cep())?,
        cnpj: required_field("cnpj", dto.cnpj())?,
    })
}
